File: solutions/ruby/diamond/2/diamond.rb
class Diamond
  def self.make_diamond(letter)
    return "A\n" if letter == "A"
    
    letters = ("A"..letter).to_a
    n = letters.length
    width = 2 * n - 1
    
    rows = []
    
    # 上半分と中段
    letters.each_with_index do |l, i|
      if i == 0
        rows << l.center(width)
      else
        row = l + (" " * (2 * i - 1)) + l
        rows << row.center(width)
      end
    end
    
    # 下半分
    letters[0...-1].reverse_each do |l|
      i = letters.index(l)
      if i == 0
        rows << l.center(width)
      else
        row = l + (" " * (2 * i - 1)) + l
        rows << row.center(width)
      end
    end
    
    rows.join("\n") + "\n"
  end
end

File: solutions/python/diamond/2/diamond.py
def rows(letter):
    if letter == 'A':
        return ['A']
    
    alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    index = alphabet.index(letter)
    width = 2 * index + 1
    
    result = []
    
    # Upper half + middle
    for i in range(index + 1):
        char = alphabet[i]
        if i == 0:
            row = ' ' * index + char + ' ' * index
        else:
            outer_spaces = ' ' * (index - i)
            inner_spaces = ' ' * (2 * i - 1)
            row = outer_spaces + char + inner_spaces + char + outer_spaces
        result.append(row)
        
    # Lower half
    for i in range(index - 1, -1, -1):
        result.append(result[i])
        
    return result
